storage/src/main/java/tech/pegasys/teku/storage/server/kvstore/KvStoreDatabase.java

```diff
@@ -1408,18 +1401,46 @@ private void removeNonCanonicalSidecars(
             .map(entry -> new SlotAndBlockRoot(entry.getValue(), entry.getKey()))
             .collect(Collectors.toSet());
 
-    LOG.trace("Removing sidecars for non-canonical blocks");
-    try (final FinalizedUpdater updater = finalizedUpdater()) {
-      for (final SlotAndBlockRoot slotAndBlockRoot : nonCanonicalBlocks) {
-        dao.getDataColumnIdentifiers(slotAndBlockRoot)
-            .forEach(
-                identifier -> {
-                  LOG.trace(
-                      "Removing sidecar with identifier {} for non-canonical block", identifier);
-                  updater.removeSidecar(identifier);
-                });
+    if (storeNonCanonicalBlocks) {
+      final Iterator<SlotAndBlockRoot> nonCanonicalBlocksIterator = nonCanonicalBlocks.iterator();
+      int index = 0;
+      while (nonCanonicalBlocksIterator.hasNext()) {
+        final int start = index;
+        try (final FinalizedUpdater updater = finalizedUpdater()) {
+          while (nonCanonicalBlocksIterator.hasNext() && (index - start) < BLOBS_TX_BATCH_SIZE) {
+            dao.getDataColumnIdentifiers(nonCanonicalBlocksIterator.next())
+                .forEach(
+                    key -> {
+                      dao.getSidecar(key)
+                          .ifPresent(
+                              sidecarBytes -> {
+                                DataColumnSidecar sideCar =
+                                    spec.deserializeSidecar(sidecarBytes, key.slot());
+                                updater.addNonCanonicalSidecar(sideCar);
+                                LOG.trace(
+                                    "Removing sidecar with identifier {} for non-canonical block",
```

Comment:
Maybe "Moving non-canonical sidecar with identifier {} to non-canonical sidecars table"